presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java

```diff
@@ -365,72 +366,86 @@ private static Expression toExpression(Object value)
+                        aliased(
+                                values(row(new NullLiteral())),
+                                "x",
+                                ImmutableList.of("no partitions")));
             }
-            TableLayout layout = getOnlyElement(layouts).getLayout();
-            if (!layout.getDiscretePredicates().isPresent()) {
-                throw new SemanticException(NOT_SUPPORTED, showPartitions, "Table does not have partition columns: %s", table);
-            }
-            List<ColumnHandle> partitionColumns = layout.getDiscretePredicates().get().getColumns();
-
-            /*
-                Generate a dynamic pivot to output one column per partition key.
-                For example, a table with two partition keys (ds, cluster_name)
-                would generate the following query:
-
-                SELECT
-                  partition_number
-                , max(CASE WHEN partition_key = 'ds' THEN partition_value END) ds
-                , max(CASE WHEN partition_key = 'cluster_name' THEN partition_value END) cluster_name
-                FROM ...
-                GROUP BY partition_number
-
-                The values are also cast to the type of the partition column.
-                The query is then wrapped to allow custom filtering and ordering.
-            */
-
-            ImmutableList.Builder<SelectItem> selectList = ImmutableList.builder();
-            ImmutableList.Builder<SelectItem> wrappedList = ImmutableList.builder();
-            selectList.add(unaliasedName("partition_number"));
-            for (ColumnHandle columnHandle : partitionColumns) {
-                ColumnMetadata column = metadata.getColumnMetadata(session, tableHandle.get(), columnHandle);
-                Expression key = equal(identifier("partition_key"), new StringLiteral(column.getName()));
-                Expression value = caseWhen(key, identifier("partition_value"));
-                value = new Cast(value, column.getType().getTypeSignature().toString());
-                Expression function = functionCall("max", value);
-                selectList.add(new SingleColumn(function, new Identifier(column.getName())));
-                wrappedList.add(unaliasedName(column.getName()));
+
+            Map<ColumnHandle, String> columnHandleToName = ImmutableBiMap.copyOf(metadata.getColumnHandles(session, tableHandle)).inverse();
+            List<String> partitioningColumns = TupleDomain.extractFixedValues(predicates.get(0)).get().keySet().stream()
+                    .map(columnHandle -> requireNonNull(columnHandleToName.get(columnHandle), "no column name for handle"))
+                    .sorted()
+                    .collect(toImmutableList());
+
+            ImmutableList.Builder<Expression> rows = ImmutableList.builder();
+            for (TupleDomain<ColumnHandle> domain : predicates) {
+                List<Expression> row = TupleDomain.extractFixedValues(domain).get()
```

Comment:
the whole conversion from `DiscretePredicates` to `TupleDomain` to fixed values is not something I'm particularly proud of, but I'm following `InformationSchemaPageSourceProvider#buildPartitions` here.